netket/optimizer/qgt/qgt_jacobian_dense.py

```diff
@@ -63,28 +63,45 @@ def QGTJacobianDense(
               models. holomorphic works for any function assuming it's holomorphic
               or real valued.
         holomorphic: a flag to indicate that the function is holomorphic.
-        diag_shift: Constant shift added to diagonal entries.
-        diag_scale: Fractional shift added to diagonal entries.
+        diag_scale: Fractional shift :math:`\epsilon_1` added to diagonal entries (see below)
+        diag_shift: Constant shift :math:`\epsilon_2` added to diagonal entries (see below)
         chunk_size: If supplied, overrides the chunk size of the variational state
                     (useful for models where the backward pass requires more
                     memory than the forward pass).
-    """
-    if mode is not None and holomorphic is not None:
-        raise ValueError("Cannot specify both `mode` and `holomorphic`.")
 
-    diag_shift, diag_scale = sanitize_diag_shift(diag_shift, diag_scale, rescale_shift)
+    Numerical estimates of the QGT are usually ill-conditioned and require
+    regularisation. The standard approach is to add a positive constant to the diagonal;
+    alternatively, Becca and Sorella (2017) propose scaling this offset with the
+    diagonal entry itself. NetKet allows using both in tandem:
+
+    .. math::
+
+        S_{ii} \mapsto S_{ii} + \epsilon_1 S_{ii} + \epsilon_2;
 
+    :math:`\epsilon_{1,2}` are specified using `diag_scale` and `diag_shift`,
+    respectively. By default, `diag_scale=0.0`, `diag_shift=0.01`.
+    """
     if vstate is None:
+        if mode is not None and holomorphic is not None:
+            raise ValueError("Cannot specify both `mode` and `holomorphic`.")
```

Comment:
this was good where it was?

Reply:
This way it will never be executed if you run `QGTJacobianDense(vstate, holomorphic=True, mode="complex")` but it should error.

Plus, I think it's better to error early if we can rather than later.

Reply:
Early because if it errors right when you specified those options it's more obvious to an user what was wrong, rather than erroring later on in the optimisation and the user has to understand where he had specified those options.

Reply:
> This way it will never be executed if you run `QGTJacobianDense(vstate, holomorphic=True, mode="complex")` but it should error.

OK, I was assuming it errors somewhere down the line, but it doesn't...

Reply:
I removed it in my commit when I moved it to the top. I noticed that we could have errored early and thought it was better.